nixos/xmrig: add kernel module msr

ratsclub opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Fixes #146416

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ratsclub committed:
nixos/xmrig: add kernel module msr

aanderse commented on nixos/modules/services/misc/xmrig.nix:
```diff
@@ -51,14 +51,15 @@ with lib;
       serviceConfig = {
         ExecStartPre = "${cfg.package}/bin/xmrig --config=${configFile} --dry-run";
         ExecStart = "${cfg.package}/bin/xmrig --config=${configFile}";
-        DynamicUser = true;
```
Can we add some capabilities instead of running as `root`?

ratsclub replied:
I tried the code below and it didn't work properly. I used [this link as a reference](https://man7.org/linux/man-pages/man7/capabilities.7.html)

```nix
systemd.services.xmrig = {
  wantedBy = [ "multi-user.target" ];
  after = [ "network.target" ];
  description = "XMRig Mining Software Service";
  serviceConfig = {
    ExecStartPre = "${cfg.package}/bin/xmrig --config=${configFile} --dry-run";
    ExecStart = "${cfg.package}/bin/xmrig --config=${configFile}";
    DynamicUser = true;
    AmbientCapabilities = "CAP_SYS_RAWIO";
  };
};
 ```

thiagokokada replied:
IMO, if someone wants to harden its system NixOS offer tools for it already (you can always add something like `systemd.services.xmrig.serviceConfig.DynamicUser = true;` directly to your config).

Unless we have some kinda of policy for service hardening, I am fine with this change.

aanderse replied:
> Unless we have some kinda of policy for service hardening

@thiagokokada shhh or else @mweinelt might hear you :eyes:

:smile:

aanderse replied:
Given this is what upstream suggests (IIRC, from what I read on their site) I'm also fine with if we can't figure out how to make it work with `DynamicUser`. Bonus points if you apply any other potential hardening, though :+1:

thiagokokada replied:
> @thiagokokada shhh or else @mweinelt might hear you eyes

I am not against hardening per see, but I think other systemd options are more effective than `DynamicUsers`, for example:

```
serviceConfig.PrivateTmp = "yes";
serviceConfig.ProtectHome = "yes";       
serviceConfig.ProtectSystem = "strict";
serviceConfig.NoNewPrivileges = "yes";
```

Etc etc.

ratsclub replied:
I added the options @thiagokokada provided to the service and it worked as expected!

mweinelt replied:
Maybe we can use `https://github.com/xmrig/xmrig/blob/dev/scripts/randomx_boost.sh` in `ExecStartPre` and make that privileged instead of the whole process?

It is mentioned here: https://xmrig.com/docs/miner/randomx-optimization-guide/msr#configuration

thiagokokada replied:
According to the [documentation](https://www.freedesktop.org/software/systemd/man/systemd.service.html#id-1.8.3.6.2.4), something like this:

```
ExecStartPre = "!${pkgs.xmrig}/scripts/randomx_boost.sh"
``` 

However, this is assuming the scripts are in the package. Also, it is probably better to patch out the `modprobe msr` part.

ratsclub replied:
I feel that putting this script is not a good idea as it has some imperative commands in it. As @thiagokokada already mentioned I also think that the hardening might cause some problems later...

thiagokokada commented on nixos/modules/services/misc/xmrig.nix:
```diff
@@ -51,14 +51,19 @@ with lib;
+        PrivateTmp = true;
+        ProtectHome = true;
+        ProtectSystem = "strict";
+        NoNewPrivileges = true;
```
Does this also works?

```suggestion
        NoNewPrivileges = true;
        PrivateDevices = true;
        LockPersonality = true;
        RestrictNamespaces = true;
        RestrictRealtime = true;
        RestrictSUIDSGID = true;
```

ratsclub replied:
I tested all options individually and removing `PrivateDevices = true;` made it work again. So I guess we only need to remove this one!

thiagokokada reviewed (commented):
BTW, I am still not sure about those hardening options. I think we should follow upstream instead. They can always cause issues and it is better for end users to set them (so they can know the consequences).

ratsclub commented:
I removed the hardening and added a comment explaining why it runs as `root`. Talking with members of the monero mining community I was informed that most of them just run the software as `sudo`, so I guess that it might be nice to leverage the responsibility to the end-user.

thiagokokada commented on nixos/modules/services/misc/xmrig.nix:
```diff
@@ -51,14 +51,20 @@ with lib;
+        # If you use recent XMRig with root privileges (Linux) or admin
+        # privileges (Windows) the miner configure all MSR registers
+        # automatically.
+        DynamicUser = false;
```
Either remove this or:

```suggestion
        DynamicUser = lib.mkDefault false;
```

Otherwise if someone tries to overwrite this they will receive an error.

thiagokokada commented:
Well, LGTM, going to merge.
